salt/pillar/__init__.py

```diff
@@ -260,40 +272,24 @@ class Pillar(object):
     '''
     def __init__(self, opts, grains, minion_id, saltenv, ext=None, functions=None,
                  pillar=None, pillarenv=None):
+
+        self.orig_opts = opts
+        self._functions = functions
         self.minion_id = minion_id
         # Store the file_roots path so we can restore later. Issue 5449
-        self.actual_file_roots = opts['file_roots']
+        self.actual_file_roots = self.orig_opts['file_roots']
         # use the local file client
-        self.opts = self.__gen_opts(opts, grains, saltenv=saltenv, ext=ext, pillarenv=pillarenv)
-        self.client = salt.fileclient.get_file_client(self.opts, True)
+        self._opts = self.__gen_opts(self.orig_opts, grains, saltenv=saltenv, ext=ext, pillarenv=pillarenv)
 
-        if opts.get('file_client', '') == 'local':
-            opts['grains'] = grains
+        # FIXME: if we change self.orig_opts then it is no longer "original".  Hmmm.
+        if self.orig_opts.get('file_client') == 'local':
+            self.orig_opts['grains'] = grains
 
-        # if we didn't pass in functions, lets load them
```

Comment:
So {{functions}} is saved in {{self._functions}}.  For future references of {{functions}} it is accessed through {{self.functions}}, a {{@property}} (line #363 green below).  The {{self.functions}} will either return the passed-in {{self._functions}} if it is non {{None}}, or it will load fresh functions.  This is necessary because the functions need the current, up-to-date pillar data, otherwise the current pillar source is unable to reference pillar data from previous pillar sources.  I did this because there wasn't an obvious way to update the dunder variables ({{__salt__}}, {{__pillar__}}) in the already-loaded functions.